content/authentication/next-steps/bonus-module/spring-security/_index.md

```diff
@@ -0,0 +1,731 @@
+
+```groovy
+    // Remove implementation("org.springframework.security:spring-security-crypto:5.5.1")
+	implementation 'org.thymeleaf.extras:thymeleaf-extras-springsecurity6'
+	implementation 'org.springframework.boot:spring-boot-starter-security'
+```
+
+This lesson will consist of four parts of implementation to adopt the
+Spring Security framework:
+
+1. Refactor `UserService` to implement Spring Security `UserDetailsService`
+interface methods that take care of authentication and user session
+1. Remove `AuthenticationFilter` and `WebApplicationConfig` to replace with
+Spring Security 6 `WebSecurity` class, which handles authorization and
+request filtering
+1. Refactor other controllers and services to make use of updated `UserService`
+1. Improve error pages and error redirection
+
+### Authentication in Spring Security 6
+
+First, it is important to take a look at the [Spring Security Authentication
+documentation](https://docs.spring.io/spring-security/reference/servlet/authentication/architecture.html),
+which describes the classes and components that implement the
+behaviors we want to use.
+
+We are going to implement username/password verification using the
+`UserDetailsService` interface.
+
+#### Implementing `UserDetailsService`
+
+The `UserDetailsService` implementation returns a `UserDetails` object which
+is a Spring Security representation of username, password, and granted user
+authorities. We are going to refactor our `UserService` class to implement
+a custom `IUserService` interface that we create. Our `IUserService` interface
+will extend the `UserDetailsService` interface that we are required to use for
+Spring Security.
+
+```mermaid{ zoom="false" }
+classDiagram
+    UserDetailsService <|-- IUserService : extends
+    IUserService <|.. UserService : implements
+    class UserDetailsService{
+        loadUserByUsername(String username) UserDetails
+    }
+    class IUserService{
+        findByUsername(String username) User
+        save(RegisterFormDTO registration) User
+        getCurrentUser() User
+        findById(Integer id) Optional~User~
+        findAll() List~User~
+        deleteUser(Integer id) User
+    }
+    class UserService{
+        loadUserByUsername(String username) UserDetails
+        save(RegisterFormDTO registration) User
+        getCurrentUser() User
+        findByUsername(String username) User
+        findById(Integer id) Optional~User~
+        findAll() List~User~
+        deleteUser(Integer id) User
+    }
+```
+
+Create a new *interface* in the `codingevents.services` package named
+`IUserService`.
+
+This interface should **extend** the `UserDetailsService` interface, which
+is found in the `org.springframework.security.core.userdetails` package.
+
+`IUserService` should include two method definitions, which mirror the
+methods we have already implemented in `UserService`.
+
+* `User findByUsername(String username);`
+* `User save(RegisterFormDTO registration);`
```

Comment:
I would encourage you to turn some of the unordered lists into ordered lists. Ordered lists are easier for screenreader users to navigate.